File: AlgorithmSelfStudy/AlgorithmC++/Baekjoon_10866.cpp
#include <iostream>

class Node {
public:
	Node* next;
	Node* prev;
	int data;
	static inline Node* insertNode(int data) {
		Node* node = new Node();
		node->data = data;
		node->next = NULL;
		node->prev = NULL;
		return node;
	}
};

class Queue {
public:
	Queue() {
		front = rear = NULL;
		size = 0;
	}
	Node* front;
	Node* rear;
	int size;
};

int main() {
	Queue* queue = new Queue();

	int n;
	scanf_s("%d", &n);
	for (int i = 0; i < n; i++) {
		std::string input;
		std::cin >> input;
		if (input == "push_front") {
			int data;
			scanf_s("%d", &data);
			Node* temp = Node::insertNode(data);
			if (queue->front == NULL)
				queue->rear = queue->front = temp;
			else {
				temp->next = queue->front;
				queue->front->prev = temp;
				queue->front = temp;
			}
			queue->size++;
		}
		else if (input == "push_back") {
			int data;
			scanf_s("%d", &data);
			Node* temp = Node::insertNode(data);
			if (queue->rear == NULL)
				queue->rear = queue->front = temp;
			else {
				temp->prev = queue->rear;
				queue->rear->next = temp;
				queue->rear = temp;
			}
			queue->size++;
		}
		else if (input == "pop_front") {
			if (queue->size == 0)
				printf("-1\n");
			else {
				printf("%d\n", queue->front->data);
				Node* temp = queue->front;
				queue->front = queue->front->next;
				free(temp);
				if (queue->front == NULL)
					queue->rear = NULL;
				else
					queue->front->prev = NULL;
				queue->size--;
			}
		}
		else if (input == "pop_back") {
			if (queue->size == 0)
				printf("-1\n");
			else {
				printf("%d\n", queue->rear->data);
				Node* temp = queue->rear;
				queue->rear = queue->rear->prev;
				if (queue->rear == NULL)
					queue->front = NULL;
				else
					queue->rear->next = NULL;
				queue->size--;
			}
		}
		else if (input == "size") {
			printf("%d\n", queue->size);
		}
		else if (input == "empty") {
			queue->size == 0 ? printf("1\n") : printf("0\n");
		}
		else if (input == "front") {
			if (queue->size == 0)
				printf("-1\n");
			else
				printf("%d\n", queue->front->data);
		}
		else if (input == "back") {
			if (queue->size == 0)
				printf("-1\n");
			else
				printf("%d\n", queue->rear->data);
		}
	}
}

File: AlgorithmSelfStudy/AlgorithmSelfStudy/STL/Baekjoon_10866.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace AlgorithmSelfStudy.STL
{
    class Baekjoon_10866
    {
        static void Main(string[] args)
        {
            int n = int.Parse(Console.ReadLine());
            int[] arr = new int[n];

            int loc = 0;
            for(int i=0; i<n; i++)
            {
                string temp = Console.ReadLine();
                string[] data = temp.Split();
                if (data[0].Equals("push_front"))
                {
                    for(int j = loc - 1; j >=0; j--)
                    {
                        arr[j + 1] = arr[j];
                    }
                    arr[0] = int.Parse(data[1]);
                    loc++;
                }
                else if (data[0].Equals("push_back"))
                {
                    arr[loc] = int.Parse(data[1]);
                    loc++;
                }
                else if (data[0].Equals("pop_front"))
                {
                    if (loc != 0)
                    {
                        Console.WriteLine(arr[0]);
                        for(int j=0; j<loc-1; j++)
                        {
                            arr[j] = arr[j + 1];
                        }
                        loc--;
                    }
                    else
                        Console.WriteLine("-1");
                }
                else if (data[0].Equals("pop_back"))
                {
                    if (loc != 0)
                    {
                        Console.WriteLine(arr[loc - 1]);
                        loc--;
                    }
                    else
                        Console.WriteLine("-1");
                }
                else if (data[0].Equals("size"))
                {
                    Console.WriteLine(loc);
                }
                else if (data[0].Equals("empty"))
                {
                    if (loc == 0)
                        Console.WriteLine("1");
                    else
                        Console.WriteLine("0");
                }
                else if (data[0].Equals("front"))
                {
                    if (loc != 0)
                        Console.WriteLine(arr[0]);
                    else
                        Console.WriteLine("-1");
                }
                else if (data[0].Equals("back"))
                {
                    if (loc != 0)
                        Console.WriteLine(arr[loc - 1]);
                    else
                        Console.WriteLine("-1");
                }
            }
        }
    }
}
